python/extensions/pip.bzl

```diff
@@ -14,63 +14,140 @@
+
+        whl_map[hub_name][whl_name][pip_attr.python_version] = pip_name + "_"
+
+    return whl_map
+
 def _pip_impl(module_ctx):
+    root_hubs = {}
+    submodules_hubs = {}
+    hub_whl_map = {}
+
+    # TODO we need to check for duplicate python versions in the same hub
+    # TODO we need to check for duplicate python hub names in submodules
+    # TODO improve failure messages
     for mod in module_ctx.modules:
-        for attr in mod.tags.parse:
-            requrements_lock = locked_requirements_label(module_ctx, attr)
-
-            # Parse the requirements file directly in starlark to get the information
-            # needed for the whl_libary declarations below. This is needed to contain
-            # the pip_repository logic to a single module extension.
-            requirements_lock_content = module_ctx.read(requrements_lock)
-            parse_result = parse_requirements(requirements_lock_content)
-            requirements = parse_result.requirements
-            extra_pip_args = attr.extra_pip_args + parse_result.options
-
-            # Create the repository where users load the `requirement` macro. Under bzlmod
-            # this does not create the install_deps() macro.
-            pip_repository_bzlmod(
-                name = attr.name,
-                repo_name = attr.name,
-                requirements_lock = attr.requirements_lock,
-                incompatible_generate_aliases = attr.incompatible_generate_aliases,
+        for pip_attr in mod.tags.parse:
+            if mod.is_root:
+                if pip_attr.hub_name in root_hubs:
+                    root_hubs[pip_attr.hub_name].append(pip_attr)
+                else:
+                    root_hubs[pip_attr.hub_name] = [pip_attr]
+            elif pip_attr.hub_name in submodules_hubs:
+                submodules_hubs[pip_attr.hub_name].append(pip_attr)
+            else:
+                submodules_hubs[pip_attr.hub_name] = [pip_attr]
+
+    for hub_name, pip_attr in root_hubs.items():
+        for pip in pip_attr:
+            # TODO we need to check to see if we have the same python version
+            # in a different pip call.
+            hub_whl_map = _create_pip(module_ctx, pip, hub_name, hub_whl_map)
+
+    for hub_name, pip_attr in submodules_hubs.items():
+        if hub_name in root_hubs:
+            # TODO clean up the warn statement like we are doing in toolchains
+            _print_warn("Not creating pip with the hub_name {}, same hub name found in the root module.".format(hub_name))
+            continue
+        for pip in pip_attr:
+            hub_whl_map = _create_pip(module_ctx, pip, hub_name, hub_whl_map)
+
+    for hub_name, whl_map in hub_whl_map.items():
+        for whl_name, version_map in whl_map.items():
+            whl_library_alias(
+                name = hub_name + "_" + whl_name,
+                wheel_name = whl_name,
+                default_version = str(DEFAULT_PYTHON_VERSION),
```

Comment:
nit: str() call isn't necessary; it's already a string.

Reply:
It is. I need to see if I can put a string in the hub, rather than an int